openrdp/__init__.py

```diff
@@ -0,0 +1,311 @@
+        aln = list(set(aln))  # Remove identical sequences
+
+        # Check that sequences are of the same length
+        seqlens = [len(s) for s in aln]
+        if len(set(seqlens)) > 1:
+            print(f"ERROR: Sequences in input {self.infile} are not the same length!")
+            sys.exit()
+
+        # check that all sequences are the same length
+        seqlen = set([len(s) for s in aln])
+        if len(seqlen) > 1:
+            print("ERROR: The input sequences are different lengths!  Input must be aligned.")
+            sys.exit()
+
+        # Create an m x n array of sequences
+        alignment = np.array(list(map(list, aln)))
+
+        # prepare return value
+        results = ScanResults(dict([(method, {}) for method in aliases.keys()]))
+
+        # Run 3Seq
+        if 'threeseq' in self.methods:
+            three_seq = ThreeSeq(self.infile)
+            self.print("Starting 3Seq Analysis")
+            results.dict['threeseq'] = three_seq.execute()
+            self.print("Finished 3Seq Analysis")
+
+        # Run GENECONV
+        if 'geneconv' in self.methods:
+            # Parse output file if available
+            if self.config:
+                geneconv = GeneConv(settings=dict(self.config.items('Geneconv')))
+            else:
+                geneconv = GeneConv()  # default config
+
+            self.print("Starting GENECONV Analysis")
+            results.dict['geneconv'] = geneconv.execute(self.infile)
+            self.print("Finished GENECONV Analysis")
+
+        # Exit early if 3Seq and Geneconv are the only methods selected
+        check = set(aliases.keys()).intersection(self.methods)
+        if len(check) == 0:
+            return results
+
+        tmethods = {}
+        for alias, a in aliases.items():
+            if alias in ['threeseq', 'geneconv'] or alias not in self.methods:
+                continue
+
+            self.print(f"Setting up {alias} analysis...")
+            if self.config:
+                settings = dict(self.config.items(a['key']))
+                tmethod = a['method'](alignment, settings=settings, quiet=self.quiet)
+            else:
+                tmethod = a['method'](alignment, quiet=self.quiet)
+            tmethods.update({alias: tmethod})
+
+        # iterate over all triplets in the alignment
+        total_num_trps = sum(1 for _ in combinations(range(alignment.shape[0]), 3))
+
+        if 'bootscan' in tmethods:
+            bootscan = tmethods['bootscan']
+            bootscan.execute_all(total_combinations=total_num_trps, seq_names=self.seq_names)
+
+        trp_count = 1
+        for trp in generate_triplets(alignment):
+            triplet = Triplet(alignment, self.seq_names, trp)
+            self.print("Scanning triplet {} / {}".format(trp_count, total_num_trps))
+            trp_count += 1
+            for alias, tmethod in tmethods.items():
+                if alias == 'bootscan':
+                    continue
+                tmethod.execute(triplet)
```

Comment:
If the `bootscan` method is being used, the `execute_all()` function generates the triplets, so there wouldn't be a need to iterate over the triplets again

Reply:
got it, pulled in your code from `scripts/run_scans.py`